Graphics/GraphicsEngineD3DBase/src/ShaderD3DBase.cpp

```diff
@@ -83,14 +224,56 @@ class D3DIncludeImpl : public ID3DInclude
     std::unordered_map<LPCVOID, RefCntAutoPtr<IDataBlob>> m_DataBlobs;
 };
 
-static HRESULT CompileShader(const char*                      Source,
-                             LPCSTR                           strFunctionName,
-                             const D3D_SHADER_MACRO*          pDefines,
-                             IShaderSourceInputStreamFactory* pIncludeStreamFactory,
-                             LPCSTR                           profile,
-                             ID3DBlob**                       ppBlobOut,
-                             ID3DBlob**                       ppCompilerOutput)
+static HRESULT CompileShader(const char*             Source,
+                             size_t                  SourceLength,
+                             const ShaderCreateInfo& ShaderCI,
+                             LPCSTR                  profile,
+                             ID3DBlob**              ppBlobOut,
+                             ID3DBlob**              ppCompilerOutput)
 {
+    std::vector<D3D_SHADER_MACRO> D3DMacros;
+    switch (ShaderCI.Desc.ShaderType)
+    {
+        case SHADER_TYPE_VERTEX:
+            D3DMacros.push_back({"VERTEX_SHADER", "1"});
+            break;
+
+        case SHADER_TYPE_PIXEL:
+            D3DMacros.push_back({"FRAGMENT_SHADER", "1"});
+            D3DMacros.push_back({"PIXEL_SHADER", "1"});
+            break;
+
+        case SHADER_TYPE_GEOMETRY:
+            D3DMacros.push_back({"GEOMETRY_SHADER", "1"});
+            break;
+
+        case SHADER_TYPE_HULL:
+            D3DMacros.push_back({"TESS_CONTROL_SHADER", "1"});
+            D3DMacros.push_back({"HULL_SHADER", "1"});
+            break;
+
+        case SHADER_TYPE_DOMAIN:
+            D3DMacros.push_back({"TESS_EVALUATION_SHADER", "1"});
+            D3DMacros.push_back({"DOMAIN_SHADER", "1"});
+            break;
+
+        case SHADER_TYPE_COMPUTE:
+            D3DMacros.push_back({"COMPUTE_SHADER", "1"});
+            break;
+
+        default: UNEXPECTED("Unexpected shader type");
+    }
+
+    if (ShaderCI.Macros)
+    {
+        for (auto* pCurrMacro = ShaderCI.Macros; pCurrMacro->Name && pCurrMacro->Definition; ++pCurrMacro)
+        {
+            D3DMacros.push_back({pCurrMacro->Name, pCurrMacro->Definition});
+        }
+    }
+
+    D3DMacros.push_back({nullptr, nullptr});
```

Comment:
Can you please move this duplicate code into a function?

Reply:
FXC uses ansi strings, DXC uses unicode strings, it is not a duplicate code